sql: enforce SELECT policies for UPDATE/DELETE

spilchen opened the pull request:
To align with postgres behavior, SELECT policies should be applied in specific cases:
- For DELETE statements, apply SELECT policies as filters.
- For UPDATE statements:
  - Apply SELECT policies as filters when selecting existing rows.
  - Apply SELECT policies as check constraints for new rows.

Additionally, when a SELECT query includes locking clauses (e.g., SELECT ... FOR UPDATE|SHARE), UPDATE policies should also be applied.

For more details on the postgres behaviour, see this: https://www.postgresql.org/docs/17/sql-createpolicy.html#SQL-CREATEPOLICY-UPDATE

Closes #136742 

Epic: CRDB-45203
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143063)
<!-- Reviewable:end -->

michae2 reviewed (changes requested):
Nice!

> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143063#-:-OLfjCr-EQxDkTLvCnY6:b-83sjoh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`-- commits` line 14 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143063#-OLfO6pFDW9UgnN63WB7:-OLfO6pFDW9UgnN63WB8:by466o):*
This issue has been deleted?
___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 2069 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143063#-OLfNF1aB9BPKDeaRmOs:-OLfNF1aB9BPKDeaRmOt:b-ww9m2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4b9742524c2d5e84e035d6c9d23f26651be2da8/pkg/sql/logictest/testdata/logic_test/row_level_security#L2069)):*
> ```
> 
> statement ok
> insert into r1 values (0),(1),(2),(3),(4),(5);
> ```

It would be interesting if r1 also had a value that didn't satisfy the upd1 policy, such as 22.
___
*[`pkg/sql/opt/optbuilder/row_level_security.go` line 252 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143063#-OLfV5K2Bmg3ged38Xi9:-OLfV5K2Bmg3ged38XiA:by3lp3u) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4b9742524c2d5e84e035d6c9d23f26651be2da8/pkg/sql/opt/optbuilder/row_level_security.go#L252)):*
> ```Go
> // combinePolicyWithCheckExpr will build a combined expression depending if this
> // is INSERT or UPDATE.
> func (r *optRLSConstraintBuilder) combinePolicyWithCheckExpr(colIDs *intsets.Fast) string {
> ```

nit: Why does `combinePolicyUsingExpressionForCommand` work with tree.TypedExprs, but `combinePolicyWithCheckExpr` work with strings?
___
*[`pkg/sql/opt/optbuilder/select.go` line 755 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143063#-OLfdVrZ3xXx4v5WPMUi:-OLfdVrZ3xXx4v5WPMUj:b-51o988) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4b9742524c2d5e84e035d6c9d23f26651be2da8/pkg/sql/opt/optbuilder/select.go#L755)):*
> ```Go
> 	// logical properties of the scan to fully normalize the index predicates.
> 	b.addPartialIndexPredicatesForTable(tabMeta, outScope.expr)
> 	b.addRowLevelSecurityFilter(tabMeta, outScope, policyCommandScope, locking)
> ```

Unfortunately, by the time we're within `buildScan`, `locking` doesn't definitively tell us whether the scan is on behalf of a SELECT FOR UPDATE or SELECT FOR SHARE statement. If we're using the new lock operator (i.e. under Read Committed isolation, or with `SET optimizer_use_lock_op_for_serializable = on;`) then `locking` could be nil here even within a SELECT FOR UPDATE statement.

I think the best way to check if this scan is for a SELECT FOR UPDATE / FOR SHARE statement would be to look at `lockCtx.locking.get().IsLocking()` just before wherever we call `b.lockingSpecForTableScan`, and then pass this boolean into `buildScan` either separately or by creating a new PolicyScope just for SFU / SFS.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql: enforce SELECT policies for UPDATE/DELETE

To align with postgres behavior, SELECT policies should be applied in
specific cases:
- For DELETE statements, apply SELECT policies as filters.
- For UPDATE statements:
  - Apply SELECT policies as filters when selecting existing rows.
  - Apply SELECT policies as check constraints for new rows.

Additionally, when a SELECT query includes locking clauses (e.g.,
SELECT ... FOR UPDATE|SHARE), UPDATE policies should also be applied.

Closes #136742

Epic: CRDB-45203
Release note: none

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143063)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[`-- commits` line 14 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143063#-OLfO6pFDW9UgnN63WB7:-OLiAUFU8cBycPI3Xzc4:b-wwsce):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

This issue has been deleted?
</blockquote></details>

Whoops. I copied the wrong one in. It should be #136742.
___
*[`pkg/sql/opt/optbuilder/row_level_security.go` line 252 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143063#-OLfV5K2Bmg3ged38Xi9:-OLiAwvdDDv5HsxVA8Xa:b-hs5jmm) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4b9742524c2d5e84e035d6c9d23f26651be2da8/pkg/sql/opt/optbuilder/row_level_security.go#L252)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

nit: Why does `combinePolicyUsingExpressionForCommand` work with tree.TypedExprs, but `combinePolicyWithCheckExpr` work with strings?
</blockquote></details>

The `WITH CHECK` expression works with strings because that is the data type used for check constraints. Ultimately, this string is returned through the `Constraint()` function of the `CheckConstraint` interface.
https://github.com/cockroachdb/cockroach/blob/8a407843cae42a5486ab8845def5be464155b32c/pkg/sql/opt/cat/table.go#L199-L207 
___
*[`pkg/sql/opt/optbuilder/select.go` line 755 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143063#-OLfdVrZ3xXx4v5WPMUi:-OLiOba1DGGL-udjd14d:bok45f0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4b9742524c2d5e84e035d6c9d23f26651be2da8/pkg/sql/opt/optbuilder/select.go#L755)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Unfortunately, by the time we're within `buildScan`, `locking` doesn't definitively tell us whether the scan is on behalf of a SELECT FOR UPDATE or SELECT FOR SHARE statement. If we're using the new lock operator (i.e. under Read Committed isolation, or with `SET optimizer_use_lock_op_for_serializable = on;`) then `locking` could be nil here even within a SELECT FOR UPDATE statement.

I think the best way to check if this scan is for a SELECT FOR UPDATE / FOR SHARE statement would be to look at `lockCtx.locking.get().IsLocking()` just before wherever we call `b.lockingSpecForTableScan`, and then pass this boolean into `buildScan` either separately or by creating a new PolicyScope just for SFU / SFS.
</blockquote></details>

Ah, okay, this makes more sense. I had mistakenly assumed that weaker isolation levels simply omitted the locking clause.

I ended up adjusting the policy scope in the callers to `b.buildScan`, rather than deep inside where the RLS filters are applied.

I also added additional testing for a scan that is over a table ref, which I didn't have before (e.g. SELECT * FROM [53 as t] FOR UPDATE).
___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 2069 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143063#-OLfNF1aB9BPKDeaRmOs:-OLiQGkeDonK_xBnURTT:bmiskcq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4b9742524c2d5e84e035d6c9d23f26651be2da8/pkg/sql/logictest/testdata/logic_test/row_level_security#L2069)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

It would be interesting if r1 also had a value that didn't satisfy the upd1 policy, such as 22.
</blockquote></details>

Yes, good idea.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice job!

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143063#-:-OLjkgGn9orF-LrHc5kn:b-46j5c8)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 2346 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143063#-OLjkbyv7ytcW1tn7aOA:-OLjkbyv7ytcW1tn7aOB:bmjvl2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bbd01ae7d1d48e8494436a77331233c6f671a3e/pkg/sql/logictest/testdata/logic_test/row_level_security#L2346)):*
> ```
> 
> statement ok
> DROP USER r1_user;
> ```

Nice test cases!


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
